django/db/migrations/state.py

```diff
@@ -98,16 +98,31 @@ def __init__(self, models=None, real_apps=None):
         self.real_apps = real_apps
         self.is_delayed = False
         # {remote_model_key: {model_key: [(field_name, field)]}}
-        self.relations = None
+        self._relations = None
+
+    @property
+    def relations(self):
+        if self._relations is None:
+            self.resolve_fields_and_relations()
+        return self._relations
 
     def add_model(self, model_state):
-        app_label, model_name = model_state.app_label, model_state.name_lower
-        self.models[(app_label, model_name)] = model_state
+        model_key = model_state.app_label, model_state.name_lower
+        self.models[model_key] = model_state
+        if self._relations is not None:
+            self.resolve_model_relations(model_key)
         if 'apps' in self.__dict__:  # hasattr would cache the property
-            self.reload_model(app_label, model_name)
+            self.reload_model(*model_key)
 
     def remove_model(self, app_label, model_name):
         del self.models[app_label, model_name]
+        if self._relations is not None:
+            removed_model_key = (app_label, model_name)
```

Comment:
Since this model key is the main model key used in this method, how about defining `model_key` in the first line of the method? Then below you can choose a different name for the model key accessed in each loop of the for loop since it's used less frequently. That would also let you change the (current) first line of the method to `del self.models[model_key]`. You could also do `unregister_model(*model_key)` towards the bottom if you wanted, like you do for `reload_model()` above.